### DISCUSSION TITLE
Is it possible to override a method of the base widget using a widget-subclass?

### DISCUSSION THREAD

**Discussion by BurningTreeC** [TiddlyWiki/TiddlyWiki5#8073] [Q&A]
I would like to override the `refreshChildren` method of the base widget (`widget.js`) using a `widget-subclass`

Is that possible somehow? I'm trying my best but with no luck...

**Comment by BurningTreeC** [comment]
I found out that I can do this in a startup module. But would that be the preferred option?

Thank you,
Simon

**Reply by pmario** [reply]
**It would be interesting, what you try to achieve?** 

From my point of view modifying the "base" widget class in a plugin is a no go, since every other widget will inherit from the base widget. Chances are very high, that there will be unwanted side effects. 

Usually we create a user-widget, which inherits from widget and then overwrite the refresh mechanism in only 1 user-widget.

**Reply by BurningTreeC** [reply]
What I'm trying to achieve is, that ALL widgets do an additional check in the `refreshChildren` method

**Reply by pmario** [reply]
How exactly should that look like. Would it make sense for the core to have this "check" built in already. So you would not need to "patch" it?

**Comment by Jermolene** [comment]
Hi @BurningTreeC it would indeed be interesting to know what you're trying to achieve. Overriding a base class does seem like it should be a last resort.

**Comment by BurningTreeC** [comment]
Hi @pmario @Jermolene 

What I'm trying to achieve is the following:

- don't refresh widgets that have a domNode that has a certain style
- in my case the style is `content-visibility: auto`

Why?

In my TiddlyFlex layout tiddlers that are outside of the visible areas get the style `content-visibility: auto` and `contain-intrinsic-size` set to the width and height of the elements. Like that, if browsers don't render the element because of `content-visibility: auto`, it still takes up its original height and width and with a resizeObserver I can detect if they resize even if they are outside of the view and can then re-calculate their dimensions or if they're visible.

Now, refreshing a tiddler with `content-visibility: auto` doesn't make sense. It makes sense doing it when its content-visibility-state changes to `visible` and I'm doing that. There are also big performance improvements possible if tiddlers with `content-visibility: auto` don't get refreshed.

How I'm doing it (and I don't say that it's the correct way) is overriding the base widget `refreshChildren` method with the following code:

```
var Widget = require("$:/core/modules/widgets/widget.js").widget;

Widget.prototype.refreshChildren = function(changedTiddlers,force) {
	var children = this.children,
		refreshed = false;
	for(var i=0; i<children.length; i++) {
		var child = children[i];
		if(!force && child.domNodes && child.domNodes[0] && child.domNodes[0].style && ((child.domNodes[0].style["content-visibility"] === "auto") || (child.domNodes[0].style["content-visibility"] === "hidden"))) {
			return false;
		} else {
			refreshed = child.refresh(changedTiddlers) || refreshed;
		}
	}
	return refreshed;
};
```

I really don't know if this is something for the core @pmario ... even if it makes sense not refreshing tiddlers with those styles ... 

Thank you for your attention!
Simon

**Reply by pmario** [reply]
>I really don't know if this is something for the core @pmario ... even if it makes sense not refreshing tiddlers with those styles ...

You already use the dynaview-plugin. So it may have a place there?

**Reply by BurningTreeC** [reply]
I use the dynaview-plugin ONLY to get the viewport dimensions in state tiddlers. For nothing else. I made a custom `dynanode` widget which does things like checking if an element is visible within a given node.

**Comment by BurningTreeC** [comment]
P.S. yes, the question is, what to do if a tiddler outside of the view transcludes a tiddler that is within the view and if that tiddler's content changes and that would cause the tiddler out of view to resize, what to do in that case... I'm testing that :smile_cat:

**Comment by BurningTreeC** [comment]
I found my answer: it's not needed. :smile:
